src/server/req_register.c

```diff
@@ -294,64 +305,79 @@ req_register(struct batch_request *preq)
 			switch (type) {
 
 				case JOB_DEPEND_TYPE_AFTERSTART:
-					if (pjob->ji_qs.ji_substate >= JOB_SUBSTATE_RUNNING) {
-						/* job already running, setup task to send	*/
-						/* release back to child and continue with	*/
-						/* registration process 			*/
-						ptask = set_task(WORK_Immed, 0, post_run_depend,
-							(void *)pjob);
-						if (ptask)
-							append_link(&pjob->ji_svrtask,
-								&ptask->wt_linkobj, ptask);
-					}
+					if (is_finished == FALSE) {
+						if (pjob->ji_qs.ji_substate >= JOB_SUBSTATE_RUNNING) {
+							/* job already running, setup task to send	*/
+							/* release back to child and continue with	*/
+							/* registration process 			*/
+							ptask = set_task(WORK_Immed, 0, post_run_depend,
+								(void *)pjob);
+							if (ptask)
+								append_link(&pjob->ji_svrtask,
+									&ptask->wt_linkobj, ptask);
+						}
+					} else
+						rc = PBSE_HISTJOBID;
 					/* fall through to complete registration */
 				case JOB_DEPEND_TYPE_AFTERANY:
 				case JOB_DEPEND_TYPE_AFTEROK:
 				case JOB_DEPEND_TYPE_AFTERNOTOK:
-					rc = register_dep(pattr, preq, type, &made);
+					/* If the job has already finished, no need to add after dependency */
+					if (is_finished == FALSE)
+						rc = register_dep(pattr, preq, type, &made);
+					else {
+						if (((type == JOB_DEPEND_TYPE_AFTERNOTOK) && (pjob->ji_qs.ji_un.ji_exect.ji_exitstat == 0)) ||
+						    ((type == JOB_DEPEND_TYPE_AFTEROK) && (pjob->ji_qs.ji_un.ji_exect.ji_exitstat != 0)))
+							rc = PBSE_HISTDEPEND;
+						else
+							rc = PBSE_HISTJOBID;
+					}
 					break;
-
 				case JOB_DEPEND_TYPE_BEFORESTART:
 				case JOB_DEPEND_TYPE_BEFOREANY:
 				case JOB_DEPEND_TYPE_BEFOREOK:
 				case JOB_DEPEND_TYPE_BEFORENOTOK:
 
-					/*
-					 * Check job owner for permission, use the real
-					 * job owner, not the sending server's name.
-					 */
-
-					(void)strcpy(preq->rq_user,
-						preq->rq_ind.rq_register.rq_owner);
-					if (svr_chk_owner(preq, pjob)) {
-						rc = PBSE_PERM;		/* not same user */
-					} else {
-						/* ok owner, see if job has "on" */
-						pdep = find_depend(JOB_DEPEND_TYPE_ON, pattr);
-						if (pdep == 0) {
-							/* on "on", see if child already registered */
-							revtype = type ^ (JOB_DEPEND_TYPE_BEFORESTART -
-								JOB_DEPEND_TYPE_AFTERSTART);
-							pdep = find_depend(revtype, pattr);
+					/* There is no need to put before dependency on a finished job */
+					if (is_finished == FALSE) {
+						/*
```

Comment:
Can you change this with
```C
if (is_finished == TRUE) {
    rc = PBSE_HISTDEPEND;
    break;
}
```
Then you don't have to mess with the indentation and the git blame.